File: fe/src/features/checkout/model/parcel.ts
import type { CartItem } from "@/shared/contracts/api/cart";
import type { ParcelDimensions } from "@/shared/contracts/api/product";

export function trustedParcelDimensions(items: readonly CartItem[]): ParcelDimensions | null {
  if (items.length === 0) {
    return null;
  }

  const aggregate: ParcelDimensions = {
    weightGrams: 0,
    lengthCm: 0,
    widthCm: 0,
    heightCm: 0,
  };

  for (const item of items) {
    const parcel = item.parcel;
    if (!parcel) {
      return null;
    }
    aggregate.weightGrams += parcel.weightGrams * item.quantity;
    aggregate.lengthCm = Math.max(aggregate.lengthCm, parcel.lengthCm);
    aggregate.widthCm = Math.max(aggregate.widthCm, parcel.widthCm);
    aggregate.heightCm = Math.max(aggregate.heightCm, parcel.heightCm);
  }

  return aggregate;
}

File: services/shipping-service/src/main/java/com/vnshop/shippingservice/domain/Parcel.java
package com.vnshop.shippingservice.domain;

public record Parcel(int weightGrams, int lengthCm, int widthCm, int heightCm, long declaredValueMinor) {
    public Parcel(int weightGrams, int lengthCm, int widthCm, int heightCm) {
        this(weightGrams, lengthCm, widthCm, heightCm, 0L);
    }
}
